### PR TITLE
Make flows deterministic

### PR DESCRIPTION
**Pull request by KillerX**
For more info see https://pkg.go.dev/go.temporal.io/sdk/contrib/tools/workflowcheck#section-readme

### CODE REVIEW

File: paths/paths.go [hunk 1]
```diff
@@ -210,6 +211,15 @@ var drivePrefixes = map[Drive]prefix{
 	TestDrive:         {"./testdata/", "./testdata/", "/dev/null/"},
 }
 
+func SafeParse(ctx workflow.Context, path string) (Path, error) {
+	var p Path
+	err := workflow.SideEffect(ctx, func(ctx workflow.Context) any {
+		return MustParse(path)
```
Comment (andreasgangso) [inline paths/paths.go:217]: MustParse panics, is that ok?

File: utils/workflows/files.go [hunk 2]
```diff
@@ -88,13 +89,63 @@ func ListFiles(ctx workflow.Context, path paths.Path) (paths.Files, error) {
...
+		return r
+	})
+
+	var data T
+	unmarsallResult.Get(data)
+
+	return &data, nil
+}
+
+func SafeMarshallXml(ctx workflow.Context, data any) ([]byte, error) {
```
Comment (andreasgangso) [inline utils/workflows/files.go:112]: Marshal* one l

File: workflows/scheduled/files_cleanup.go [hunk 3]
```diff
@@ -21,9 +21,14 @@ func CleanupTemp(ctx workflow.Context) (CleanupResult, error) {
 	ctx = workflow.WithActivityOptions(ctx, wfutils.GetDefaultActivityOptions())
 
 	deletedFiles := []string{}
-	err := wfutils.ExecuteWithLowPrioQueue(ctx, activities.Util.DeleteOldFiles, activities.CleanupInput{
-		Root:      paths.MustParse("/mnt/temp/"),
-		OlderThan: time.Now().Add(-14 * 24 * time.Hour),
+	rootPath, err := paths.SafeParse(ctx, "/mnt/temp/")
```
Comment (andreasgangso) [inline workflows/scheduled/files_cleanup.go:24]: I liked MustParse for paths like this

File: workflows/scheduled/files_cleanup.go [hunk 4]
```diff
@@ -38,8 +43,9 @@ func CleanupTemp(ctx workflow.Context) (CleanupResult, error) {
 		return res, err
 	}
 
+	temp, _ := paths.SafeParse(ctx, "/mnt/temp/")
```
Comment (andreasgangso) [inline workflows/scheduled/files_cleanup.go:46]: same path as line 24

File: paths/paths.go [hunk 5]
```diff
@@ -210,6 +211,15 @@ var drivePrefixes = map[Drive]prefix{
 	TestDrive:         {"./testdata/", "./testdata/", "/dev/null/"},
 }
 
+func SafeParse(ctx workflow.Context, path string) (Path, error) {
```
Comment (andreasgangso) [inline paths/paths.go:214]: Maybe this shouldnt be in paths, but rather in wfutils or another place.
Or rename ParseSideEffect() or something like that.

File: utils/workflows/files.go [hunk 6]
```diff
@@ -88,13 +89,63 @@ func ListFiles(ctx workflow.Context, path paths.Path) (paths.Files, error) {
...
+	var res []byte
+	err := workflow.SideEffect(ctx, func(ctx workflow.Context) any {
```
Comment (andreasgangso) [inline utils/workflows/files.go:114]: feel `safe` is misdescriptive.
Maybe have a package called `sideeffect`: sideeffect.marshalXml(), sideeffect.parsePath(), etc.
Or wfutils.SEMarshalXML() 
Or simply wfutils.MarshalXML()

**Review by andreasgangso — APPROVED** [review]

**Comment by andreasgangso** [comment]
@KillerX can you look at this before lunch tomorrow? id like to merge #250 but id like to use the marshal thingies from here

File: paths/paths.go (hunk 1 above)
Comment (KillerX) [reply paths/paths.go:217]: Yes. Basically the idea is that if the workflow is deterministic then the parameters will always be the same so retrying with the same values is useless, thus a panic which is a "non-retryable error" in essence. The panic is caught by temporal so it does not affect the other threads/routines.

**Comment by KillerX** [comment]
I basically reverted the `SafePath` thing now and marked `MustParse` deterministic, as the output is independent of the iteration order.
Also fixed names on the `Marshal` functions.

**Pull request merged**